pkg/analyzer/nodes/node.go

```diff
@@ -103,22 +104,45 @@ func extractFilesFromDirectory(root string, matcher matcher.PrefixMatcher) (*fil
 	m := metrics.FileExtractionMetrics{}
 	// TODO(ROX-13771): Use `range matcher.GetCommonPrefixDirs()` again after fixing.
 	for _, dir := range []string{"etc/", "usr/share/rpm", "var/lib/rpm", "usr/share/buildinfo"} {
-		if err := n.addFiles(filepath.FromSlash(dir), matcher, &m); err != nil {
+		if err := n.addFiles(ctx, filepath.FromSlash(dir), matcher, &m); err != nil {
 			return nil, errors.Wrapf(err, "failed to match filesMap at %q (at %q)", dir, n.root)
 		}
 	}
 	m.Emit()
 	return n, nil
 }
 
+func walkDirWithContext(ctx context.Context, dir string, fn fs.WalkDirFunc) error {
+	errC := make(chan error)
+	go func(ctx context.Context) {
+		defer close(errC)
+		errC <- filepath.WalkDir(dir, func(path string, info fs.DirEntry, err error) error {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				return fn(path, info, err)
+			}
```

Comment:
You can avoid one more WalkDir() iteration and potentially one more file system call if you run `fn()` first, and check the context after. Most of the time, the context will be canceled while `fn()` is running. Good to avoid another call because if we are timing out, that call might hang for a while and you would have this goroutine running for longer unnecessarily.

```suggestion
			if fErr := fn(path, info, err); fErr != nil {
				return fErr
			}
			select {
			case <-ctx.Done():
				return ctx.Err()
			default:
				return nil
			}
```

Reply:
Got that covered! One more change was needed to prevent a goroutine leak.